### PR TITLE
AI Bundle

### PR DESCRIPTION
**Pull request by grammeaway**

### CODE REVIEW

File: RacingGame/AI/src/main/java/racing/ai/services/AIPlugin.java [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+import racing.common.map.MapSPI;
+import racing.common.map.Tile;
+import racing.common.services.IGamePluginService;
+import racing.common.ai.AISPI;
+import racing.common.data.TileType;
+import racing.common.data.entityparts.TilePart;
+
+/**
+ *
+ * @author Victor Gram & Niclas Johansen
```
Comment (dueruen) [inline RacingGame/AI/src/main/java/racing/ai/services/AIPlugin.java]: Should not be here

File: RacingGame/AI/src/main/java/racing/ai/services/AIPlugin.java [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+ */
+public class AIPlugin implements IGamePluginService, AISPI {
```
Comment (dueruen) [inline RacingGame/AI/src/main/java/racing/ai/services/AIPlugin.java]: No reason to implement IGamePluginService. It is not used

File: RacingGame/AI/src/main/java/racing/ai/services/AIProcessingSystem.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+import racing.common.data.GameData;
+import racing.common.data.World;
+import racing.common.services.IEntityProcessingService;
+
+/**
+ *
+ * @author Victor Gram & Niclas Johansen
+ */
+public class AIProcessingSystem implements IEntityProcessingService {
```
Comment (dueruen) [inline RacingGame/AI/src/main/java/racing/ai/services/AIProcessingSystem.java]: Class not used

File: RacingGame/AI/src/main/resources/META-INF/entityprocessor.xml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (dueruen) [inline RacingGame/AI/src/main/resources/META-INF/entityprocessor.xml]: Can be remove if AIProcessingSystem.java is remove

File: RacingGame/AI/src/main/resources/META-INF/gameplugin.xml [hunk 5]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<scr:component xmlns:scr="http://www.osgi.org/xmlns/scr/v1.1.0" name="racing.ai.services.plugin">
+    <implementation class="racing.ai.services.AIPlugin"/>
+    <service>
+        <provide interface="racing.common.services.IGamePluginService"/>
```
Comment (dueruen) [inline RacingGame/AI/src/main/resources/META-INF/gameplugin.xml]: Is line can be remove if AIPlugin no longer implments IGamePluginService

File: RacingGame/CommonAI/bin/osgi.bnd [hunk 6]
```diff
@@ -0,0 +1 @@
+Export-Package: racing.common.ai
```
Comment (dueruen) [inline RacingGame/CommonAI/bin/osgi.bnd]: Bin folder is not needed

File: RacingGame/CommonAI/bin/pom.xml [hunk 7]
```diff
@@ -0,0 +1,46 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (dueruen) [inline RacingGame/CommonAI/bin/pom.xml]: Bin folder is not needed

File: RacingGame/CommonAI/src/main/java/racing/common/ai/AISPI.java [hunk 8]
```diff
@@ -0,0 +1,16 @@
+package racing.common.ai;
+
+import java.util.ArrayList;
+import racing.common.data.Entity;
+import racing.common.data.World;
+import racing.common.data.entityparts.PositionPart;
+
+/**
+ *
+ * @author Victor Gram & Niclas Johansen
```
Comment (dueruen) [inline RacingGame/CommonAI/src/main/java/racing/common/ai/AISPI.java]: Remove

File: RacingGame/CommonNPC/bin/osgi.bnd [hunk 9]
```diff
@@ -0,0 +1,5 @@
+<<<<<<< HEAD
```
Comment (dueruen) [inline RacingGame/CommonNPC/bin/osgi.bnd]: Bin folder is not needed

File: RacingGame/CommonNPC/bin/pom.xml [hunk 10]
```diff
@@ -0,0 +1,46 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (dueruen) [inline RacingGame/CommonNPC/bin/pom.xml]: Bin folder is not needed

File: RacingGame/CommonNPC/osgi.bnd [hunk 11]
```diff
@@ -1 +1,5 @@
-Export-Package: racing.common.npc
```
Comment (dueruen) [inline RacingGame/CommonNPC/osgi.bnd:1]: Fix

File: RacingGame/CommonNPC/src/main/java/racing/common/npc/NPCSPI.java [hunk 12]
```diff
@@ -4,9 +4,12 @@
 import racing.common.data.World;
 
 /**
- * Npc spi
+ *
+ * @author Victor Gram & Niclas Johansen
```
Comment (dueruen) [inline RacingGame/CommonNPC/src/main/java/racing/common/npc/NPCSPI.java]: remove

File: RacingGame/NPC/bin/pom.xml [hunk 13]
```diff
@@ -32,13 +32,20 @@
             <version>${project.version}</version>
```
Comment (dueruen) [inline RacingGame/NPC/bin/pom.xml]: Bin folder is not needed

File: RacingGame/NPC/bin/src/main/resources/META-INF/entityprocessor.xml [hunk 14]
```diff
@@ -4,4 +4,7 @@
     <service>
```
Comment (dueruen) [inline RacingGame/NPC/bin/src/main/resources/META-INF/entityprocessor.xml]: Bin folder is not needed

File: RacingGame/NPC/src/main/java/racing/npc/NPCPlugin.java [hunk 15]
```diff
@@ -4,13 +4,18 @@
...
 import racing.common.data.entityparts.MovingPart;
 import racing.common.data.entityparts.PositionPart;
 import racing.common.data.entityparts.ScorePart;
 import racing.common.npc.NPC;
 import racing.common.npc.NPCSPI;
 import racing.common.services.IGamePluginService;
 
+/**
+ *
+ * @author Victor Gram & Niclas Johansen
```
Comment (dueruen) [inline RacingGame/NPC/src/main/java/racing/npc/NPCPlugin.java]: remove

File: RacingGame/NPC/src/main/java/racing/npc/NPCProcessingSystem.java [hunk 16]
```diff
@@ -1,64 +1,248 @@
...
+import racing.common.data.entityparts.ItemPart;
+import racing.common.data.entityparts.TilePart;
+import racing.common.map.MapSPI;
+import racing.common.map.Tile;
 import racing.common.npc.NPC;
 
 /**
- *Processing system for NPC entity
+ *
+ * @author Victor Gram & Niclas Johansen
```
Comment (dueruen) [inline RacingGame/NPC/src/main/java/racing/npc/NPCProcessingSystem.java]: remove

**Review by dueruen — CHANGES_REQUESTED** [review]

File: RacingGame/AI/src/main/java/racing/ai/services/AIPlugin.java (hunk 2 above)
Comment (grammeaway) [reply RacingGame/AI/src/main/java/racing/ai/services/AIPlugin.java]: The start and stop methods both come from IGamePluginService, and it also makes it possible for the OSGi framework to load the bundle as a pluginservice, so I think it's needed - correct me if I'm wrong

Comment (dueruen) [reply RacingGame/AI/src/main/java/racing/ai/services/AIPlugin.java]: By removing IGamePluginService the class still implements AISPI and the bundle will be loaded

File: RacingGame/AI/src/main/java/racing/ai/astar/AStar.java [hunk 17]
```diff
@@ -0,0 +1,244 @@
...
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.PriorityQueue;
+import java.util.Set;
+
+/**
+ * @author Marcelo Surriabre
+ * https://github.com/marcelo-s/A-Star-Java-Implementation
+ * modifications by Victor Gram & Niclas Johansen
```
Comment (dueruen) [inline RacingGame/AI/src/main/java/racing/ai/astar/AStar.java]: Or maybe Group 09

File: RacingGame/AI/src/main/java/racing/ai/astar/AStarNode.java [hunk 18]
```diff
@@ -0,0 +1,124 @@
+package racing.ai.astar;
+
+/**
+ * @author Marcelo Surriabre
+ * https://github.com/marcelo-s/A-Star-Java-Implementation
+ * modifications by Victor Gram & Niclas Johansen
```
Comment (dueruen) [inline RacingGame/AI/src/main/java/racing/ai/astar/AStarNode.java]: Or maybe Group 09

File: RacingGame/AI/src/main/java/racing/ai/services/AIPlugin.java [hunk 19]
```diff
@@ -0,0 +1,208 @@
...
+     * Declarative service remove map service
+     *
+     * @param map map service
+     */
+    public void removeMapService(MapSPI map) {
+        this.mapSPI = null;
+    }
+
+    @Override
+    public void start(GameData gameData, World world) {
```
Comment (dueruen) [inline RacingGame/AI/src/main/java/racing/ai/services/AIPlugin.java]: remove

File: RacingGame/AI/src/main/java/racing/ai/services/AIPlugin.java [hunk 20]
```diff
@@ -0,0 +1,208 @@
...
+    }
+
+    @Override
+    public void stop(GameData gameData, World world) {
```
Comment (dueruen) [inline RacingGame/AI/src/main/java/racing/ai/services/AIPlugin.java]: remove

File: RacingGame/AI/src/main/java/racing/ai/services/AIPlugin.java [hunk 21]
```diff
@@ -0,0 +1,208 @@
...
+import racing.common.map.Tile;
+import racing.common.services.IGamePluginService;
+import racing.common.ai.AISPI;
+import racing.common.data.TileType;
+import racing.common.data.entityparts.TilePart;
+
+/**
+ * Plugin handling the addition and removal of the plugin from the system
+ */
+public class AIPlugin implements IGamePluginService, AISPI {
```
Comment (dueruen) [inline RacingGame/AI/src/main/java/racing/ai/services/AIPlugin.java]: remove IGamePluginService

**Review by dueruen — CHANGES_REQUESTED** [review]

**Review by dueruen — APPROVED** [review]

**Pull request merged**